File: FrontEnd/lib/domain/repositories/i_post_repository.dart
/// @file i_post_repository.dart
/// @description Interfaz del repositorio de posts de la comunidad.
/// Define el contrato que debe cumplir cualquier implementación de repositorio de posts.
/// @module Community
/// @layer Domain
library;

import '../dtos/community/create_post_request_dto.dart';
import '../entities/comment.dart';
import '../entities/post.dart';

/// Contrato del repositorio de posts y comentarios de la comunidad.
///
/// Los use cases dependen de esta interfaz para desacoplarse de la
/// implementación concreta (PostRepositoryImpl).
abstract interface class IPostRepository {

  /// Obtiene la página [page] del feed de la comunidad.
  ///
  /// [param] page     — Número de página (1-based).
  /// [param] limit    — Posts por página.
  /// [param] authorId — Si se proporciona, llama a /community/mine (posts propios).
  /// [returns] Lista de posts del feed.
  /// [throws] AppError — Si la petición falla.
  Future<List<Post>> getFeed({int page = 1, int limit = 20, String? authorId});

  /// Obtiene un post por su identificador.
  ///
  /// [param] postId — Identificador del post.
  /// [returns] Post completo con todos sus campos.
  /// [throws] AppError — Si el post no existe.
  Future<Post> getPostById(String postId);

  /// Crea un nuevo post en la comunidad.
  ///
  /// [param] dto — DTO con los datos del nuevo post.
  /// [returns] Post recién creado.
  /// [throws] AppError — Si la validación falla o la petición da error.
  Future<Post> createPost(CreatePostRequestDto dto);

  /// Da like a un post. Operación idempotente.
  ///
  /// [param] postId — Identificador del post.
  /// [throws] AppError — Si la petición falla.
  Future<void> likePost(String postId);

  /// Quita el like de un post. Operación idempotente.
  ///
  /// [param] postId — Identificador del post.
  /// [throws] AppError — Si la petición falla.
  Future<void> unlikePost(String postId);

  /// Obtiene los comentarios de un post.
  ///
  /// [param] postId — Identificador del post.
  /// [returns] Lista de comentarios ordenados por fecha de creación.
  /// [throws] AppError — Si la petición falla.
  Future<List<Comment>> getComments(String postId);

  /// Crea un comentario en un post.
  ///
  /// [param] postId  — Identificador del post.
  /// [param] content — Texto del comentario.
  /// [returns] Comentario recién creado.
  /// [throws] AppError — Si la petición falla.
  Future<Comment> createComment(String postId, String content);
}

File: BackEnd/src/domain/repositories/IPostRepository.ts
/**
 * @file IPostRepository.ts
 * @description Interfaz del repositorio de posts de la comunidad. Define el contrato de
 * acceso a datos sin acoplar la capa de dominio a ninguna implementación concreta.
 * @module Community
 * @layer Domain
 */

import type { ClientSession } from 'mongodb';
import type { Post } from '../entities/Post.js';

/**
 * Contrato del repositorio de posts.
 * Los use cases dependen de esta interfaz, nunca de la implementación concreta.
 */
export interface IPostRepository {
  /**
   * Obtiene el feed global de posts activos, ordenado por fecha descendente.
   *
   * @param page — Número de página (base 1).
   * @param limit — Elementos por página.
   * @returns Lista paginada de posts.
   */
  findFeed(page: number, limit: number): Promise<Post[]>;

  /**
   * Busca un post activo por su id.
   *
   * @param id — Id del post.
   * @returns Post encontrado o null.
   */
  findById(id: string): Promise<Post | null>;

  /**
   * Crea un nuevo post en la base de datos.
   *
   * @param post — Datos del post sin id.
   * @param session — Sesión de transacción opcional.
   * @returns Post creado con id asignado.
   */
  create(post: Omit<Post, 'id' | 'isActive'>, session?: ClientSession): Promise<Post>;

  /**
   * Incrementa o decrementa el contador de likes de un post de forma atómica.
   *
   * @param id — Id del post.
   * @param delta — +1 para like, -1 para unlike.
   * @param session — Sesión de transacción opcional.
   */
  incrementLikesCount(id: string, delta: 1 | -1, session?: ClientSession): Promise<void>;

  /**
   * Incrementa o decrementa el contador de comentarios de un post de forma atómica.
   *
   * @param id — Id del post.
   * @param delta — +1 al añadir, -1 al borrar.
   * @param session — Sesión de transacción opcional.
   */
  incrementCommentsCount(id: string, delta: 1 | -1, session?: ClientSession): Promise<void>;

  /**
   * Soft-delete de un post individual.
   * Marca deletedAt con la fecha actual.
   *
   * @param id — ID del post a eliminar.
   */
  softDelete(id: string): Promise<void>;

  /**
   * Soft-delete masivo de todos los posts activos de un usuario.
   * Usado al eliminar la cuenta con preserveContent=false.
   *
   * @param userId — ID del usuario cuyos posts se eliminan.
   */
  softDeleteByUserId(userId: string): Promise<void>;
}
